tower-util/src/combinators/map_err.rs

```diff
@@ -0,0 +1,72 @@
+use futures_util::{future::MapErr as MapErrFut, TryFutureExt};
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+#[derive(Debug)]
+pub struct MapErr<S, F> {
+    inner: S,
+    f: F,
+}
+
+impl<S, F> MapErr<S, F> {
+    pub fn new(inner: S, f: F) -> Self {
+        MapErr { f, inner }
+    }
+}
+
+impl<S, F, Request, Error> Service<Request> for MapErr<S, F>
+where
+    S: Service<Request>,
+    F: FnOnce(S::Error) -> Error,
+    F: Clone,
```

Comment:
Using `FnOnce + Clone` here seemed a little weird at first, but then I realized that the requirement that the function be moved into the `MapErr` future and cloned in `Layer::layer` kind of drives this. So if we're going to have to require `Clone`, it makes sense to use `FnOnce`.